File: src/api/src/domain/Yoma.Core.Domain/MyOpportunity/Models/MyOpportunity.cs
using Newtonsoft.Json;
using Yoma.Core.Domain.BlobProvider;
using Yoma.Core.Domain.Entity;
using Yoma.Core.Domain.Lookups.Models;

namespace Yoma.Core.Domain.MyOpportunity.Models
{
  public class MyOpportunity
  {
    public Guid Id { get; set; }

    public Guid UserId { get; set; }

    public string UserEmail { get; set; }

    public string UserDisplayName { get; set; }

    public DateTimeOffset? UserDateOfBirth { get; set; }

    public string? UserGender { get; set; }

    public Guid? UserCountryId { get; set; }

    public string? UserCountry { get; set; }

    public string? UserEducation { get; set; }

    public Guid? UserPhotoId { get; set; }

    [JsonIgnore]
    public StorageType? UserPhotoStorageType { get; set; }

    [JsonIgnore]
    public string? UserPhotoKey { get; set; }

    public string? UserPhotoURL { get; set; }

    public Guid OpportunityId { get; set; }

    public string OpportunityTitle { get; set; }

    public string OpportunityDescription { get; set; }

    public string OpportunityType { get; set; }

    public Guid OpportunityStatusId { get; set; }

    public Opportunity.Status OpportunityStatus { get; set; }

    public string OpportunityCommitmentIntervalDescription { get; set; }

    #region Verification Counts
    public int OpportunityParticipantCountTotal { get; set; }
    #endregion Verification Counts

    public DateTimeOffset OpportunityDateStart { get; set; }

    public DateTimeOffset? OpportunityDateEnd { get; set; }

    public bool OpportunityCredentialIssuanceEnabled { get; set; }

    public string? OpportunitySSISchemaName { get; set; }

    public Guid OrganizationId { get; set; }

    public string OrganizationName { get; set; }

    public Guid? OrganizationLogoId { get; set; }

    [JsonIgnore]
    public StorageType? OrganizationLogoStorageType { get; set; }

    [JsonIgnore]
    public string? OrganizationLogoKey { get; set; }

    public string? OrganizationLogoURL { get; set; }

    public Guid OrganizationStatusId { get; set; }

    public OrganizationStatus OrganizationStatus { get; set; }

    public Guid ActionId { get; set; }

    public Action Action { get; set; }

    public Guid? VerificationStatusId { get; set; }

    public VerificationStatus? VerificationStatus { get; set; }

    public string? CommentVerification { get; set; }

    public DateTimeOffset? DateStart { get; set; }

    public DateTimeOffset? DateEnd { get; set; }

    public DateTimeOffset? DateCompleted { get; set; }

    public decimal? ZltoReward { get; set; }

    public decimal? YomaReward { get; set; }

    public DateTimeOffset DateCreated { get; set; }

    public DateTimeOffset DateModified { get; set; }

    public List<MyOpportunityVerification>? Verifications { get; set; }

    public List<Skill>? Skills { get; set; }
  }
}

File: src/web/src/api/models/myOpportunity.ts
import type {
  ErrorResponseItem,
  FormFile,
  Geometry,
  PaginationFilter,
} from "./common";
import type { VerificationType } from "./opportunity";

export interface MyOpportunityRequestVerify {
  certificate: FormFile | null;
  voiceNote: FormFile | null;
  picture: FormFile | null;
  geometry: Geometry | null;
  dateStart: string | null;
  dateEnd: string | null;
}

export enum VerificationStatus {
  None,
  Pending,
  Rejected,
  Completed,
}

export enum Action {
  Viewed,
  Saved,
  Verification,
}

export interface MyOpportunitySearchFilter
  extends MyOpportunitySearchFilterBase {}

export interface MyOpportunitySearchFilterBase extends PaginationFilter {
  action: Action;
  verificationStatuses: VerificationStatus[] | null;
}

export interface MyOpportunitySearchFilterAdmin
  extends MyOpportunitySearchFilterBase {
  userId: string | null;
  opportunity: string | null;
  organizations: string[] | null;
  valueContains: string | null;
}

export interface MyOpportunitySearchResults {
  totalCount: number | null;
  items: MyOpportunityInfo[];
}

export interface MyOpportunityInfo {
  id: string;
  userId: string;
  userEmail: string;
  userDisplayName: string | null;
  userCountry: string | null;
  userEducation: string | null;
  userPhotoId: string | null;
  userPhotoURL: string | null;
  opportunityId: string;
  opportunityTitle: string;
  opportunityDescription: string;
  opportunityType: string;
  opportunityCommitmentIntervalDescription: string;
  opportunityParticipantCountTotal: number;
  opportunityDateStart: string;
  opportunityDateEnd: string | null;
  organizationId: string;
  organizationName: string;
  organizationLogoURL: string | null;
  actionId: string;
  action: Action;
  verificationStatusId: string | null;
  verificationStatus: VerificationStatus | null | string; //NB
  commentVerification: string | null;
  dateStart: string | null;
  dateEnd: string | null;
  dateCompleted: string | null;
  dateModified: string | null; //TODO: add to api
  zltoReward: number | null;
  yomaReward: number | null;
  verifications: MyOpportunityInfoVerification[] | null;
  skills: Skill[] | null;
}

export interface Skill {
  id: string;
  name: string;
  infoURL: string | null;
}

export interface MyOpportunityInfoVerification {
  verificationType: VerificationType | string; //NB
  geometry: Geometry | null;
  fileId: string | null;
  fileURL: string | null;
}

export interface MyOpportunityRequestVerifyFinalize {
  opportunityId: string;
  userId: string;
  status: VerificationStatus;
  comment: string;
}
export interface MyOpportunityRequestVerifyFinalizeBatch {
  status: VerificationStatus;
  comment: string;
  items: MyOpportunityRequestVerifyFinalizeBatchItem[];
}

export interface MyOpportunityRequestVerifyFinalizeBatchItem {
  opportunityId: string;
  userId: string;
}

export interface MyOpportunityResponseVerifyFinalizeBatch {
  items: MyOpportunityResponseVerifyFinalizeBatchItem[];
  status: VerificationStatus | string;
}

export interface MyOpportunityResponseVerifyFinalizeBatchItem {
  opportunityId: string;
  opportunityTitle: string;
  userId: string;
  userDisplayName: string | null;
  success: boolean;
  failure: ErrorResponseItem | null;
}

export interface MyOpportunityResponseVerify {
  status: VerificationStatus | string; //NB
  comment: string | null;
}

export interface MyOpportunitySearchCriteriaOpportunity {
  id: string;
  title: string;
}
